c/tskit/trees.c

```diff
@@ -2636,12 +2639,573 @@ check_sites(const tsk_id_t *sites, tsk_size_t num_sites, tsk_size_t num_site_row
+
+    if (num_positions == 0) {
+        return ret; // No need to verify positions if there aren't any
+    }
+
+    for (i = 0; i < num_positions - 1; i++) {
+        if (positions[i] < 0 || positions[i] >= sequence_length) {
+            ret = TSK_ERR_POSITION_OUT_OF_BOUNDS;
+            goto out;
+        }
+        if (positions[i] > positions[i + 1]) {
+            ret = TSK_ERR_STAT_UNSORTED_POSITIONS;
+            goto out;
+        }
+        if (positions[i] == positions[i + 1]) {
+            ret = TSK_ERR_STAT_DUPLICATE_POSITIONS;
+            goto out;
+        }
+    }
+    // check bounds of last value
+    if (positions[i] < 0 || positions[i] >= sequence_length) {
+        ret = TSK_ERR_POSITION_OUT_OF_BOUNDS;
+        goto out;
+    }
+out:
+    return ret;
+}
+
+static int
+positions_to_tree_indexes(const tsk_treeseq_t *ts, const double *positions,
+    tsk_size_t num_positions, tsk_id_t **tree_indexes)
+{
+    int ret = 0;
+    tsk_id_t tree_index = 0;
+    tsk_size_t i, num_trees = ts->num_trees;
+
+    // This is tricky. If there are 0 positions, we calloc a size of 1
+    // we must calloc, because memset will have no effect when called with size 0
+    *tree_indexes = tsk_calloc(num_positions, sizeof(*tree_indexes));
+    if (tree_indexes == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+    tsk_memset(*tree_indexes, TSK_NULL, num_positions * sizeof(**tree_indexes));
+    for (i = 0; i < num_positions; i++) {
+        while (ts->breakpoints[tree_index + 1] <= positions[i]) {
+            tree_index++;
+        }
+        (*tree_indexes)[i] = tree_index;
+    }
+    tsk_bug_assert(tree_index <= (tsk_id_t)(num_trees - 1));
+
+out:
+    return ret;
+}
+
+static int
+get_index_counts(
+    const tsk_id_t *indexes, tsk_size_t num_indexes, tsk_size_t **out_counts)
+{
+    int ret = 0;
+    tsk_id_t index = indexes[0];
+    tsk_size_t count, i;
+    tsk_size_t *counts = tsk_calloc(
+        (tsk_size_t)(indexes[num_indexes ? num_indexes - 1 : 0] - indexes[0] + 1),
+        sizeof(*counts));
+    if (counts == NULL) {
+        ret = TSK_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    count = 1;
+    for (i = 1; i < num_indexes; i++) {
+        if (indexes[i] == indexes[i - 1]) {
+            count++;
+        } else {
+            counts[index - indexes[0]] = count;
+            count = 1;
+            index = indexes[i];
+        }
+    }
+    counts[index - indexes[0]] = count;
+    *out_counts = counts;
+out:
+    return ret;
+}
+
+typedef struct {
+    tsk_tree_t *tree;
+    tsk_bit_array_t *node_samples;
+    tsk_id_t *parent;
+    tsk_id_t *edges_out;
+    tsk_id_t *edges_in;
+    double *branch_len;
+    tsk_size_t n_edges_out;
+    tsk_size_t n_edges_in;
+} iter_state;
+
+static int
+iter_state_init(iter_state *self, const tsk_treeseq_t *ts, tsk_size_t state_dim)
+{
+    int ret = 0;
+    const tsk_size_t num_nodes = ts->tables->nodes.num_rows;
+
+    self->tree = tsk_malloc(sizeof(*self->tree));
```

Comment:
Hmm, we also need to memset the struct itself to ensure all the pointers are zero. Here's how I would do it:
```
memset(self, 0, sizeof(*self));

ret = tsk_tree_init(&self->tree, ts, TSK_NO_SAMPLE_COUNTS);
```

By using an embedded tree struct (rather than a pointer like you currently have) we avoid having to do the malloc, and (more importantly) making sure that the memory is zero'd out before any possible error conditions can occur.

Reply:
I like the direction of using an embedded pointer. I've incorporated that change.

I think the one thing I'm on the fence about is the use of `memset` inside of `iter_state_init`. I'm not sure it provides much benefit (please correct me if I'm wrong) because we will still need to do the `memset` of `l_state` and `r_state` in the calling function (`tsk_treeseq_two_branch_count_stat`). This is because if we fail to initialize `l_state`, `r_state` must be zero'd at the time of freeing, which is done in the calling function. A refactor would still need to preserve this behavior unless the initialization of `l_state` and `r_state` were somehow decomposed (which I don't see a strong case for because they're both needed in the subsequent for loops).

I'd be happy to add it in for safety, I just don't see what it adds exactly.

Reply:
Sure, it's not that important. I wrote the above before seeing the calling code